docs/proposals/20210524-managed-external-etcd.md

```diff
@@ -0,0 +1,570 @@
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+### Abbreviations:
+- CAPI: Cluster API
+- KCP: KubeadmControlPlane
+- CRD: Custom Resource Definition
+
+## Summary
+
+This proposal aims to extend CAPI to support provisioning and managing of the lifecycle of external etcd clusters. With this, users can create CAPI clusters with the external etcd topology where the etcd cluster is also managed by CAPI.
+
+## Motivation
+
+- Cluster API supports the use of an external etcd cluster, by allowing users to provide their external etcd cluster's endpoints.
+So it would be good to add support for provisioning the etcd cluster too, so the user can have a single workflow for full management of a cluster using the external etcd topology. The user can offload the responsibility of managing an external etcd cluster to CAPI.
+    
+- External etcd topology decouples the control plane and etcd member. So if a control plane-only node fails, or if there's a memory leak in a component like kube-apiserver, it won't directly impact an etcd member.
+  
+- Etcd is resource intensive, so it is safer to have dedicated nodes for etcd, it could use more disk space, higher bandwidth. Having a separate etcd cluster for these reasons could ensure a more resilient HA setup.
+
+### Goals
+
+- Define a contract for adding pluggable etcd providers
+  - Define the responsibilities of an etcd provider, along with the required services and API types it should add.
+  - Extend the contract for control plane providers and infrastructure providers in order to properly handle external etcd clusters managed by pluggable etcd providers.
+  - User should be able to create and manage a Kubernetes cluster that uses external etcd topology in a declarative way. The CAPI controllers will deploy the etcd cluster first and then use that as input to the workload cluster, without requiring any user intervention.
+  - There will be a 1:1 mapping between the external etcd cluster and the workload cluster.
+  - An etcd provider should support etcd clusters of any size (odd numbers only), including single member etcd clusters for testing purposes. But the documentation should specify the recommended cluster size for production use cases to range between 3 and 7.
+- Provide a first implementation of this pluggable etcd provider using [etcdadm](https://github.com/kubernetes-sigs/etcdadm).
+  - The etcdadm based provider will utilize the existing Machine objects to represent etcd members for convenience instead of adding a new machine type for etcd.
+  - Support the following etcd cluster management actions: scale up and scale down, etcd member replacement and etcd version upgrades.
+
+
+### Non-Goals
+- The first iteration will use IP addresses/hostnames as etcd cluster endpoints. It can not configure static endpoint(s) till the Load Balancer provider is available.
```

Comment:
you're right that using IP address adds a risk of control plane rollouts when an etcd member changes. However we can control the number of control plane rollouts by pausing the KCP till the etcd cluster upgrade is completed. That way the KCP nodes will only rollout once.
It would be ideal to replace this with static endpoints, either through the load balancer provider when it's available, or using kube-vip